File: leetgo/cpp/LCR 149.cong-shang-dao-xia-da-yin-er-cha-shu-lcof/solution.cpp
// Created by k at 2024/05/13 21:34
// leetgo: dev
// https://leetcode.cn/problems/cong-shang-dao-xia-da-yin-er-cha-shu-lcof/

#include <bits/stdc++.h>
#include "LC_IO.h"
using namespace std;

// @lc code=begin

class Solution {
  public:
    vector<int> decorateRecord(TreeNode *root)
    {
        vector<int> ans;
        if (root == nullptr) {
            return ans;
        }
        vector<TreeNode *> queue = {root};
        while (!queue.empty()) {
            TreeNode *node = queue.front();
            ans.push_back(node->val);
            if (node->left != nullptr) {
                queue.push_back(node->left);
            }
            if (node->right != nullptr) {
                queue.push_back(node->right);
            }
            queue.erase(queue.cbegin());
        }
        return ans;
    }
};

// @lc code=end

int main()
{
    ios_base::sync_with_stdio(false);
    stringstream out_stream;

    TreeNode *root;
    LeetCodeIO::scan(cin, root);

    Solution *obj = new Solution();
    auto res      = obj->decorateRecord(root);
    LeetCodeIO::print(out_stream, res);
    cout << "\noutput: " << out_stream.rdbuf() << endl;

    delete obj;
    return 0;
}

File: leetgo/go/LCR 149.cong-shang-dao-xia-da-yin-er-cha-shu-lcof/solution.go
// Created by k at 2024/05/13 15:36
// leetgo: dev
// https://leetcode.cn/problems/cong-shang-dao-xia-da-yin-er-cha-shu-lcof/

package main

import (
	"bufio"
	"fmt"
	"os"

	. "github.com/j178/leetgo/testutils/go"
)

// @lc code=begin

func decorateRecord(root *TreeNode) (ans []int) {
	if root == nil {
		return []int{}
	}
	queue := []*TreeNode{root}
	for len(queue) != 0 {
		node := queue[0]
		queue = queue[1:]
		ans = append(ans, node.Val)
		if node.Left != nil {
			queue = append(queue, node.Left)
		}
		if node.Right != nil {
			queue = append(queue, node.Right)
		}
	}
	return
}

// @lc code=end

func main() {
	stdin := bufio.NewReader(os.Stdin)
	root := Deserialize[*TreeNode](ReadLine(stdin))
	ans := decorateRecord(root)

	fmt.Println("\noutput:", Serialize(ans))
}
